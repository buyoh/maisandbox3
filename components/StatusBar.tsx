import React from 'react';
import StatusBadge from './StatusBadge';

type ClickHandler = (key: string) => void;
type Key = string;
type Item = {
  color: string,
  text: string,
  key: Key,
  onClick?: ClickHandler
};

type StatusBarProps = {
  values: Array<Item>,
  active?: Key
}

type StatusBarState = {
}

class StatusBar extends React.Component<StatusBarProps, StatusBarState> {
  constructor(props: StatusBarProps) {
    super(props);
    this.generateJSXFromItem = this.generateJSXFromItem.bind(this);
  }

  private generateJSXFromItem(item: Item): JSX.Element {
    return (
      <div className="flex_elem_fix" key={item.key} onClick={() => { item.onClick?.call(null, item.key); }}>
        <StatusBadge color={item.color} active={item.key === this.props.active}>{item.text}</StatusBadge>
      </div>
    );
  }

  render(): JSX.Element {
    return (
      <div className="flex_row" style={{
        borderTop: '2px solid #111',
        borderBottom: '2px solid #666',
        padding: '2px',
        overflow: 'hidden'
      }}>
        {this.props.values?.map(this.generateJSXFromItem)}
      </div>
    );
  }
}

export default StatusBar;