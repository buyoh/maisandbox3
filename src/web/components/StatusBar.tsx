import React from 'react';
import StatusBadge from './StatusBadge';

type ClickHandler = (key: string) => void;
type Key = string;
export type Item = {
  color: string;
  text: string;
  key: Key;
  onClick?: ClickHandler;
};

type StatusBarProps = {
  values: Array<Item>;
  active?: Key;
};

type StatusBarState = {};

class StatusBar extends React.Component<StatusBarProps, StatusBarState> {
  constructor(props: StatusBarProps) {
    super(props);
    this.generateJSXFromItem = this.generateJSXFromItem.bind(this);
  }

  private generateJSXFromItem(item: Item): JSX.Element {
    return (
      <div
        className="flex-elem-fix"
        key={item.key}
        onClick={() => {
          item.onClick?.call(null, item.key);
        }}
      >
        <StatusBadge
          color={item.color}
          active={item.key === this.props.active}
          clickable={!!item.onClick}
        >
          {item.text}
        </StatusBadge>
      </div>
    );
  }

  render(): JSX.Element {
    // TODO: StatusBar の高さが一定でない
    // TODO: ステータスの数が溢れた場合の考慮が無い
    return (
      <div
        className="flex-row"
        style={{
          borderTop: '2px solid #111',
          borderBottom: '2px solid #666',
          padding: '2px',
          overflow: 'hidden',
        }}
      >
        {this.props.values?.map(this.generateJSXFromItem)}

        <div
          // Status が空であっても一定の高さを保つためのダミー
          className="flex-elem-fix"
          key={'_tail'}
        >
          <StatusBadge color={'gray'} active={false} clickable={false}>
            {'>_<'}
          </StatusBadge>
        </div>
      </div>
    );
  }
}

export default StatusBar;
