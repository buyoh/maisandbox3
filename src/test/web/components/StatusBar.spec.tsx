import React from 'react';
// import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import StatusBar, { Item } from '../../../web/components/StatusBar';

test('StatusBar', () => {
  const values: Item[] = [
    {
      color: 'gray',
      text: 'zero',
      key: 'key-0',
    },
    {
      color: 'success',
      text: 'one',
      key: 'key-1',
    },
  ];

  {
    const { asFragment } = render(
      <StatusBar values={values} active={'key-0'} />
    );
    const rendered = asFragment();
    expect(rendered).toMatchSnapshot();
  }
  {
    const { asFragment } = render(
      <StatusBar values={values} active={'key-1'} />
    );
    const rendered = asFragment();
    expect(rendered).toMatchSnapshot();
  }
});
