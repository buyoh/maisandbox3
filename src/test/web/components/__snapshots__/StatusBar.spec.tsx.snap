// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBar 1`] = `
<div
  className="flex-row"
  style={
    Object {
      "borderBottom": "2px solid #666",
      "borderTop": "2px solid #111",
      "overflow": "hidden",
      "padding": "2px",
    }
  }
>
  <div
    className="flex-elem-fix"
    onClick={[Function]}
  >
    <div
      className="color-gray  "
    >
      zero
    </div>
  </div>
  <div
    className="flex-elem-fix"
    onClick={[Function]}
  >
    <div
      className="color-success "
    >
      one
    </div>
  </div>
</div>
`;

exports[`StatusBar 2`] = `
<div
  className="flex-row"
  style={
    Object {
      "borderBottom": "2px solid #666",
      "borderTop": "2px solid #111",
      "overflow": "hidden",
      "padding": "2px",
    }
  }
>
  <div
    className="flex-elem-fix"
    onClick={[Function]}
  >
    <div
      className="color-gray "
    >
      zero
    </div>
  </div>
  <div
    className="flex-elem-fix"
    onClick={[Function]}
  >
    <div
      className="color-success  "
    >
      one
    </div>
  </div>
</div>
`;
